### PR TITLE
GH Action to release the container image

### LINKED ISSUE #503: Create GitHub Actions workflow to publish Operator container image
### Description

As we are no longer able to use Quay auto-builds to publish the container due to the changes in tags we need a GitHub Actions workflow that can build and publish the Operator container image when new tags are created.

Similar workflows where created for keycloak and keycloak-containers:

* https://github.com/keycloak/keycloak/pull/10683
* https://github.com/keycloak/keycloak-containers/pull/385

### Discussion

_No response_

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by andreaTP**
## Related GH Issue

Resolves #503 

## Additional Information

Tested publishing the image to my private docker.io account:
https://hub.docker.com/repository/docker/andreatp/keycloak-operator

### CODE REVIEW

File: .github/workflows/release-container.yml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  jobs:
+    publish:
+      if: github.repository == 'keycloak/keycloak-operator'
+      runs-on: ubuntu-latest
+      steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v1
```
Comment (vmuzikar) [inline .github/workflows/release-container.yml:19]: We don't need QEMU.

**Review by vmuzikar — CHANGES_REQUESTED** [review]
@andreaTP Can't access the test build in your private repo but I trust you it went ok. 😜

**Comment by vmuzikar** [comment]
@andreaTP I presume you did try deploying the operator using the image built by this workflow, right?

File: .github/workflows/release-container.yml [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        with:
+          registry: quay.io
+          username: ${{ secrets.QUAY_USERNAME }}
+          password: ${{ secrets.QUAY_PASSWORD }}
+
+      - name: Build and push
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          platforms: linux/amd64,linux/arm64
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply .github/workflows/release-container.yml]: removed

File: .github/workflows/release-container.yml (hunk 1 above)
Comment (andreaTP) [reply .github/workflows/release-container.yml]: removed

**Comment by andreaTP** [comment]
> @andreaTP I presume you did try deploying the operator using the image built by this workflow, right?

yup, checked 👍

**Review by vmuzikar — CHANGES_REQUESTED** [review]
I'm probably doing something wrong but I'm not successful at testing it. Steps I did:
1. `make cluster/prepare`.
2. Update [`operator.yaml`](https://github.com/keycloak/keycloak-operator/blob/main/deploy/operator.yaml) with your image: `andreatp/keycloak-operator:6.7.8`.
3. `kubectl apply -f operator.yaml -n keycloak`

Getting the following error:
```
          Error response from daemon: OCI runtime create failed: container_linux.go:380:
          starting container process caused: exec: "keycloak-operator": executable
          file not found in $PATH: unknown
```

**Comment by andreaTP** [comment]
@vmuzikar `6.7.8` is the new operator, tested with `2.3.4` (still running):
Screenshot 2022-03-14 at 09 21 06

**Comment by vmuzikar** [comment]
@andreaTP Ok then. I presumed all tags are the old operator, just saw arm at the 2.3.4 build so assumed it's older.

**Review by vmuzikar — APPROVED** [review]
LGTM, but I'd like @stianst to check this too before merging.

File: .github/workflows/release-container.yml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+          username: ${{ secrets.QUAY_USERNAME }}
+          password: ${{ secrets.QUAY_PASSWORD }}
+
+      - name: Build and push
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          platforms: linux/amd64
+          push: ${{ github.event_name != 'pull_request' }}
+          tags: ${{ steps.meta.outputs.tags }}
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply .github/workflows/release-container.yml:44]: done at line 25:
https://github.com/keycloak/keycloak-operator/pull/504/files#diff-69a4610cce35bf32d579cc8d14aa59de81d80d23f0269d80861bceb81f57b176R25

[the remark this answers is not in the document]
Comment (andreaTP) [reply .github/workflows/release-container.yml:44]: What about `latest` ?

[the remark this answers is not in the document]
Comment (andreaTP) [reply .github/workflows/release-container.yml:47]: Pretty original encoding, but done:
Screenshot 2022-03-14 at 10 54 53

**Review by vmuzikar — APPROVED** [review]
LGTM, but let's wait for @stianst on this one.

File: .github/workflows/release-container.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (stianst) [reply .github/workflows/release-container.yml:47]: Take a look at https://github.com/keycloak/keycloak-containers/pull/385/files#diff-69a4610cce35bf32d579cc8d14aa59de81d80d23f0269d80861bceb81f57b176R36-R38

That just uses the raw types to create "18.0.0-legacy" and "legacy" tags. With the semver it also would tag as `latest` and potentially overwriting the new operator.

[the remark this answers is not in the document]
Comment (andreaTP) [reply .github/workflows/release-container.yml:47]: @stianst if you really prefer the encoding linked I can change this, but, here `latest` is explicitly disabled here:
https://github.com/keycloak/keycloak-operator/pull/504/files#diff-69a4610cce35bf32d579cc8d14aa59de81d80d23f0269d80861bceb81f57b176R27

and I tested that's a correct encoding and it doesn't overwrite `latest`

**Review by stianst — APPROVED** [review]

**Pull request merged**